File: 1019-next-greater-node-in-linked-list/1019-next-greater-node-in-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nextLargerNodes(ListNode* head) {
        stack<vector<int>> mono;
        map<int, int> next;
        int i = 0;
        while(head!=nullptr){
            while(!mono.empty()&&head->val>mono.top()[1]){
                next[mono.top()[0]]=head->val;
                mono.pop();
            }
            vector<int> val = {i, head->val};
            mono.push(val);
            i++;
            head=head->next;
        }
        while(!mono.empty()){
            next[mono.top()[0]]=0;
            mono.pop();
        }
        vector<int> nextLarger(i);
        for(auto h: next)   nextLarger[h.first]=h.second;
        return nextLarger;
    }
};

File: 1019-next-greater-node-in-linked-list/1019-next-greater-node-in-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nextLargerNodes(self, head: Optional[ListNode]) -> List[int]:
        answer = []
        ptr = head
        mono_stack = []
        next_ = {}
        i = 0
        while ptr:
            while len(mono_stack) > 0 and ptr.val > mono_stack[-1][1]:
                next_[mono_stack.pop()] = ptr.val
            
            mono_stack.append( (i,ptr.val) )
            
            i+=1
            ptr = ptr.next
            
        while len(mono_stack)>0:
            next_[mono_stack.pop()] = 0
            
        answer = []
        ptr = head
        i = 0
        while ptr:
            answer.append(next_[(i, ptr.val)])
            i+=1
            ptr = ptr.next
        return answer
                